command/runner.go

```diff
@@ -45,26 +59,33 @@ func (args *Args) buildCommandInput(command pulumi.StringInput, env pulumi.Strin
 		return strings.Join(inputs, " ")
 	}).(pulumi.StringOutput)
 
-	var prefix string
-	if sudo {
-		prefix = "sudo"
-	}
+	commandEscaped := command.ToStringOutput().ApplyT(func(command string) string {
+		return shellescape.Quote(command)
+	}).(pulumi.StringOutput)
+
+	return pulumi.Sprintf("%s %s bash -c %s", prefix, envVarsStr, commandEscaped)
+}
 
-	return pulumi.Sprintf("%s %s %s", prefix, envVarsStr, command)
+type runnerConfiguration struct {
+	user       string
+	connection remote.ConnectionInput
 }
 
 type Runner struct {
 	e           config.CommonEnvironment
 	namer       namer.Namer
-	connection  remote.ConnectionInput
 	waitCommand *remote.Command
+	config      runnerConfiguration
 }
 
-func NewRunner(e config.CommonEnvironment, connName string, conn remote.ConnectionInput, readyFunc func(*Runner) (*remote.Command, error)) (*Runner, error) {
+func NewRunner(e config.CommonEnvironment, connName, asUser string, conn remote.ConnectionInput, readyFunc func(*Runner) (*remote.Command, error)) (*Runner, error) {
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
What about using the functional options pattern ? Or taking a config struct instead ? I don't like that we need to pass `asUser` even when we want a runner using current user, it can be confusing.

Alternatively, you could create a NewRunnerAsUser with the extra param in the signature

Reply:
Great suggestions! I did not know about functional options pattern. I will look into it.